Packages: Add react-i18n

sirreal opened the pull request:
#### Changes proposed in this Pull Request

Add react functionality for `@wordpress/i18n`:
- `I18nContext` - localized content should be wrapped in `<I18nContext.Provider locale={ 'en' } onLocalChange={ getNextLocalData } />`.
- `useI18n` - provides translation functions `__`, `_n`, `_nx`, and `_x` which update when the context changes.
- `withI18n` HOC providing the same translation functions

Currently provides no functionality for updating the translations in `@wordpress/i18n` via `setLocaleData`.

`@wordpress/i18n` translation functions do not consider React. A reliable way to make sure translations are rerendered when locale changes is necessary. Largely outlined by https://github.com/WordPress/gutenberg/issues/19210 

https://github.com/Automattic/wp-calypso/pull/39410 provides a handy way to demo the functionality in Gutenboarding.

#### Testing instructions

Demo in #39410

sirreal committed:
Add react-i18n package

sirreal committed:
Fix tsconfig lib

sirreal committed:
Add to monorepo

p-jackson commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,58 @@
+export const withI18n = ( WrappedComponent: React.ComponentType ): React.ComponentType => {
+	return class extends React.Component {
+		render() {
+			const i18n = useI18n();
```
Given this is a class component maybe this should be using `I18nContext.Consumer`? Unless we can use context in class components now, if so then TIL!

aduth replied:
Alternatively, why does it need to be a class component, vs.

```js
return ( props ) => <WrappedComponent { ...useI18n() } { ...props } />;
```

ockham commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,58 @@
+ * import { useI18n } from '@automattic/react-i18n';
+ * function MyComponent() {
+ *   const { __ } = useI18n();
+ *   return <div>{ __( 'Translate me.' ) }</div>;
```
Maybe add translation domain to the example?

```suggestion
 *   return <div>{ __( 'Translate me.', 'textdomain' ) }</div>;
```

ockham commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,58 @@
+ *
+ * import { withI18n } from '@automattic/react-i18n';
+ * function MyComponent( { __ } ) {
+ *   return <div>{ __( 'Translate me.' ) }</div>;
```
Same as https://github.com/Automattic/wp-calypso/pull/39388/files#r378087650

```suggestion
 *   return <div>{ __( 'Translate me.', 'textdomain' ) }</div>;
```

ockham reviewed (approved):
Loving it! Tested through #39389; left two very minor notes.

Feel free to :ship: after addressing Phillip's and Andrew's notes!

sirreal committed:
Add text-domain to examples

sirreal committed:
Simplify withI18n HOC

sirreal committed:
Improve types

sirreal committed:
Add slug to hook to trigger rerender

sirreal committed:
Update package-lock

This is a partial commit of updated package-lock. Babel changes have
been removed because they've demonstrated issues recently.

simison commented on package-lock.json:
```diff
@@ -37845,8 +37852,7 @@
 				"@babel/runtime": "^7.8.3",
 				"debug": "^4.1.1",
-				"qs": "^6.5.2",
-				"wpcom-xhr-request": "^1.1.2"
```
How come no more `wpcom-xhr-request` in this PR?

sgomes replied:
Probably because `package-lock.json` hadn't been updated.

This is an expected change, since this package was never really used (we use `wpcom-proxy-request` instead of this in Calypso). See #39142.

sirreal replied:
Right, this seems to be part of a stale-ish lockfile. I left babel package movement out, but I expect this to be safe.

sirreal commented on packages/react-i18n/package.json:
```diff
@@ -0,0 +1,41 @@
+	},
+	"peerDependencies": {
+		"@wordpress/i18n": "<=3",
+		"react": ">=16.8"
```
Hooks version.

blowery replied:
might want to make this `^16.8.0` instead. We don't really know if we'll be compatible with 17+ at this point.

sirreal replied:
How do we reason about this?

- Expanding versions can be a patch update?
- Restricting versions is breaking (and means the package is already broken)?

sirreal replied:
In that case, `@wordpress/i18n@<=3` might have been good 🤔

sirreal commented on packages/react-i18n/package.json:
```diff
@@ -0,0 +1,41 @@
+		"tslib": "^1.10.0"
+	},
+	"peerDependencies": {
+		"@wordpress/i18n": "<=3",
```
This should likely become `*` 🤔 

```suggestion
		"@wordpress/i18n": "*",
```

simison commented:
A brief `README.md` could be good to add even if just with _"React bindings for @wordpress/i18n"_; I find them nice when browsing the code in Github.

sirreal committed:
Relax `@wordpress/i18n` peerDep

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+import { __, _n, _nx, _x } from '@wordpress/i18n';
+
+export const I18nContext = React.createContext< undefined | string >( undefined );
```
Can we make the default value be `'en'`? Even if you use `undefined`, it effectively means `en`, because the strings in the source are left unchanged, i.e., in English.

sirreal replied:
🤔 I'm not sure. Even Calypso has a configurable default locale:

https://github.com/Automattic/wp-calypso/blob/5a5fb58300b6b19848992517fffc36fc4e4b4fb9/config/_shared.json#L23

jsnajdr replied:
Yeah, but I think that config value is not very useful and I'd like to eventually remove it.

Think about it this way: the default language is the one that the strings in the source code are written it:
```js
__( 'Stats and Insights' )
```
These strings will be displayed in the UI when the `__` function is identity. And they are always in _some_ language, it's never anything that has `undefined` or `null` meaning.

The default language is the language that the sources are written it. It can theoretically be non-`en`, but I don't think that any part of our i18n toolchain supports that.

simison replied:
Do folks using wp-i18n ever use string keys? E.g. `__( 'stats.heading' )` gets translated to `'Stats and Insights'` when locale is English.

aduth replied:
@simison Last I recall, there's no localization data assigned for the default locale, so it will always use whatever's passed as the string argument in those cases (tl;dr: No).

sirreal replied:
> the default language is the one that the strings in the source code are written it

Sure, that makes sense. But it's convention that it would be English, I could write my source in any language and provide English via translations if I desired. `undefined` seems like a good way of saying "do not translate." Another option might be a fixed string like `"default"`.

jsnajdr replied:
> But it's convention that it would be English, I could write my source in any language and provide English via translations if I desired.

Then, if you write your source in Spanish, for example, then the locale in the context is supposed to be `es`, not `undefined`. The UI, if not translated, is in Spanish. The value of the locale can be used also by components like `<MomentProvider>` to format dates in Spanish. `undefined` is not a locale name that's helpful for moment.js.

`undefined` is meaningful in the case where message IDs are used, as @simison mentioned. `__( 'stats.heading' )` doesn't produce anything usable until a translation file is provided. Message IDs are common on some platforms (Win32 or Java comes to mind), but WordPress doesn't use them.

I'd advise strongly to use `en` as a default, because that's in fact the WordPress default.

sirreal commented:
> A brief `README.md` could be good to add even if just with _"React bindings for @wordpress/i18n"_; I find them nice when browsing the code in Github.

I promise to do this in an imminent follow up if we're happy to ship the code here 😉

simison commented:
>Agreed. I promise to do this in an imminent follow up if we're happy to ship the code here 

Haha, deal! I wanted to mention something about tests too but I won't. 🙈

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,68 @@
+export const useI18n = (): I18nProps => {
+	const locale = React.useContext( I18nContext );
+	const [ i18n, setI18n ] = React.useState< I18nProps >( makeI18n( locale ) );
+	React.useEffect( () => setI18n( makeI18n( locale ) ), [ locale ] );
```
You shouldn't need effects and local state here at all. If the locale in context changes, the whole tree should be rerendered in one pass, triggered by the context change.

With this code, if the provider is rerendered with new value `cs` instead of the old `en`, all components that use `useI18n` will be rerendered (`useContext` will trigger that). The returned `i18n` object will still be the old one (state wasn't updated yet), its `i18nLocale` field will be `en`, but `__()` will be returning `cs` translations, because the global `@wordpress/i18n` has already loaded them!

Then the state updates triggered by `useEffect` will start to fire, individually for each component. Each of them will be rerendered again, with the updated `i18n` object, where the `i18nLocale` field is finally `cs`.

That's quite messy, isn't it? 🙂 

I recommend storing the whole `{ __, _x, ..., locale }` object in the context, not just the locale string. Then the `useI18n` hook becomes trivial:
```js
const useI18n = () => useContext( I18nContext );
```
and all the logic is in the provider. The provider will be a wrapper around the context provider, maintaining a local derived state (using `useMemo`) mapping the locale string to the `i18n` object.

The [localized-moment](https://github.com/Automattic/wp-calypso/tree/master/client/components/localized-moment) package is a good blueprint for this. The `<MomentProvider>` is much more complicated than we need here, because it also handles async loading of the locale data. When the provider gets rerendered from:
```jsx
<MomentProvider currentLocale="en" />
```
to:
```jsx
<MomentProvider currentLocale="cs" />
```
the context update and rerendering of the subtree doesn't happen immediately, but only after the new locale is loaded from the network (as a dynamic webpack chunk).

sirreal replied:
👍 Good feeback here, thank you.

> The `<MomentProvider>` is much more complicated than we need here, because it also handles async loading of the locale data. When the provider gets rerendered from:

That's very helpful! I expect the provider to become very similar to to the `MomentProvider`. We need a way to fetch new locales and update the translated strings.

I was planning to make the provider accept an async function that resolves with updated locales, where the provider would handle updating `@wordpress/i18n` locales:

```tsx
<I18nProvider locale={ currentLocale } onLocaleChange={ fetchLocalesFromNetwork } />
```

In Calypso, `fetchLocalesFromNetwork` would contain something analagous to `switchLocale`:

https://github.com/Automattic/wp-calypso/blob/5a5fb58300b6b19848992517fffc36fc4e4b4fb9/client/lib/i18n-utils/switch-locale.js#L111-L154

But other projects could customize behavior.

sirreal replied:
I've implemented what's outlined in 98a6dd0

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,68 @@
+ */
+export const withI18n = < T extends I18nProps = I18nProps >(
+	WrappedComponent: React.ComponentType< T >
+): React.FunctionComponent< Optionalize< T, I18nProps > > => props => {
```
These types are awesome 😉

sirreal replied:
I'll have to share credit, I came across an organization of TypeScript cookbooks: https://github.com/typescript-cheatsheets/react-typescript-cheatsheet

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,68 @@
+): React.FunctionComponent< Optionalize< T, I18nProps > > => props => {
+	return (
+		<WrappedComponent
+			{ ...useI18n() }
```
nit: it's more common to call hooks in the component function's "prelude", if you know what I mean, instead of being inline like this. Here the hook is still always called, so it's just a style nit.

sirreal committed:
Un-inline hook

sirreal committed:
Rework context

sirreal committed:
Unique i18n instances 🎉

sirreal committed:
Revert "Unique i18n instances 🎉"

This reverts commit 75fce4cd59f6fb5296811cd4d3180936ef280d7e.

sirreal commented:
@jsnajdr I had a bnit of fun in 75fce4c which largely implements the unique `@wordpress/i18n` instances as outlined in https://github.com/WordPress/gutenberg/issues/19210 by leveraging require cache. 

Touching the require cache is probably pretty risky, so I've reverted.

sirreal committed:
Tighten peerDependency ranges

yansern reviewed (approved):
I did the `changeLocale()` test through the other PR https://github.com/Automattic/wp-calypso/pull/39410 and it works!

sirreal commented:
Thanks for all the feedback and testing everyone 🙇 

I'm going to land this as a first iteration, but I'm happy to continue addressing any further feedback in follow-up PRs.

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -0,0 +1,102 @@
+	/**
+	 * A callback that resolves with the translations data
+	 */
+	onLocaleChange?( newLocale: string | undefined ): Promise< object >;
```
The `onLocaleChange` name can be confusing. Typically, the `locale` and `onLocaleChange` pair would be used for a controlled component where an user interaction (or other external event) can change the value:
```js
<input type="locale" locale={ this.state.locale } onLocaleChange={ locale => this.setState( { locale } ) } />
```

But this function does something very different. It implements an incomplete interface. Maybe `loadLocaleData` would be a better name?

sirreal commented:
Implemented @jsnajdr's reccomendations in https://github.com/Automattic/wp-calypso/pull/39441
